### PR TITLE
Remove the zero index from VarZeroVec

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes https://github.com/unicode-org/icu4x/issues/5516

Also removes `METADATA_WIDTH`, we've set it to 0 and aren't using it. It's an unnecessary untested complication on our unsafe code.

### CODE REVIEW

**Comment by Manishearth** [comment]
I still need to run all the datagens and fix non-`zerovec` tests, but this works within zv.

**Comment by sffc** [comment]
> Also removes `METADATA_WIDTH`, we've set it to 0 and aren't using it. It's an unnecessary untested complication on our unsafe code.

It was there to enable us to use `FlexZeroVec` as the index store. But, making VZV cheaper overall means that we can lean toward designs with more, shorter VZVs with Index8, instead of fewer, longer ones with Index16. We also now have ZeroTrie which covers some of these types of cases. I'm okay removing `METADATA_WIDTH` so long as we keep the door open to potentially re-adding a more compact index structure in the future.

**Comment by Manishearth** [comment]
> It was there to enable us to use FlexZeroVec as the index store.

Yeah, that would require far more changes to the unsafe code of VZV than just METADATA_WIDTH, METADATA_WIDTH didn't particularly set the stage for anythig there code-wise.

It would be interesting encapsulating more and more of the format into the trait to allow for fancier custom formats.

File: utils/zerovec/src/varzerovec/components.rs [hunk 1]
```diff
@@ -529,58 +552,77 @@ where
...
-    fn binary_search_impl(
+        if range.start > self.len() {
+            return None;
+        }
+        if range.end > self.len() {
+            return None;
+        }
+        if range.end < range.start {
+            return None;
+        }
```
Comment (sffc) [inline utils/zerovec/src/varzerovec/components.rs:574]: Nit: You need only 2 of these 3 conditions; the third is implied

File: utils/zerovec/src/varzerovec/components.rs [hunk 2]
```diff
@@ -529,58 +552,77 @@ where
...
+            size = end - start;
+            // This establishes mid < end (which implies mid < len)
+            // size is end - start. start + size is end (which is <= len).
+            // mid = start + size/2 will be less than end
+            let mid = start + size / 2;
+
+            // Safety: mid is < end <= len, so in-range
+            let cmp = predicate(self.get_unchecked(mid));
+
+            match cmp {
```
Comment (sffc) [inline utils/zerovec/src/varzerovec/components.rs:609]: Question: why did you expand the binary_search impl instead of using the std one?

**Review by sffc — COMMENTED** [review]
To components.rs

File: utils/zerovec/src/varzerovec/components.rs (hunk 2 above)
Comment (Manishearth) [reply utils/zerovec/src/varzerovec/components.rs:609]: The hacky impl was written assuming a slice to search on. Now I could still use a byte slice that I have cut out appropriately, but the indices slice that we currently use is no longer long enough for this to work, and it was going to get more complicated, so I decided to dispense with the weird hack. I can bring it back and make it work with the weird hack if desired.

**Review by sffc — APPROVED** [review]

**Review by robertbastian — APPROVED** [review]

**Pull request merged**